### PR TITLE
Adding initial project stucture

### PR DESCRIPTION
**Pull request by ssbarnea**

### CODE REVIEW

**Comment by webknjaz** [comment]
Sorry, I haven't found time to post comments, I'll do so post-merge. I've seen a number of occurences of hardcoded mentions of other projects and unnecessary stuff.

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+skip_install = true
+# Ref: https://twitter.com/di_codes/status/1044358639081975813
+commands =
+  # build wheel and sdist using PEP-517
+  {envpython} -c 'import os.path, shutil, sys; \
+  dist_dir = os.path.join("{toxinidir}", "dist"); \
+  os.path.isdir(dist_dir) or sys.exit(0); \
+  print("Removing \{!s\} contents...".format(dist_dir), file=sys.stderr); \
+  shutil.rmtree(dist_dir)'
+  {envpython} -m pep517.build \
```
Comment (webknjaz) [inline tox.ini:103]: The CLI was experimental, is deprecated now, and will be removed in the future. You should use pypa/build instead.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    --source \
+    --binary \
```
Comment (webknjaz) [inline tox.ini:105]: When you migrate to `build`, drop these because starting with v0.5.0, it has a special logic to build wheels from sdists which is preferable for testing reproducibility.

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  REQUESTS_CA_BUNDLE  # https proxies
+  SSL_CERT_FILE  # https proxies
+  LANG
+  LC_ALL
+  LC_CTYPE
+# recreate = True
+setenv =
+  COVERAGE_FILE = {env:COVERAGE_FILE:{toxworkdir}/.coverage.{envname}}
+  PIP_DISABLE_PIP_VERSION_CHECK = 1
+  PIP_CONSTRAINT = {toxinidir}/constraints.txt
```
Comment (webknjaz) [inline tox.ini:54]: It's a good idea to have this when building the dists so that sdist/wheel are reproducible (because this would force pinned PEP 517 build env)

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+skip_missing_interpreters = True
+# `usedevelop = true` overrides `skip_install` instruction, it's unwanted
+usedevelop = false
+
+[testenv]
+description =
+  Run the tests under {basepython} and
+  devel: ansible devel branch
+  ansible29: ansible 2.9
+  core: ansible-base 2.10
```
Comment (webknjaz) [inline tox.ini:17]: core!=base

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  py: ansible-core 2.11
+extras =
+  test
+deps =
+  ansible29: ansible>=2.9,<2.10
+  py: ansible-core>=2.11
```
Comment (webknjaz) [inline tox.ini:23]: no "base"?

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  devel: ansible-core @ git+https://github.com/ansible/ansible.git  # GPLv3+
+commands =
+  # safety measure to assure we do not accidentally run tests with broken dependencies
+  {envpython} -m pip check
```
Comment (webknjaz) [inline tox.ini:27]: I'd put this into `commands_pre =`

File: test/test_api.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+"""Tests for ansiblecompat package."""
+
+
+def test_placeholder() -> None:
+    """Placeholder test."""
```
Comment (webknjaz) [inline test/test_api.py:5]: I normally also add `assert True` to such placeholder tests. But it's mostly a matter of taste.

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#! /usr/bin/env python3
+"""Ansible-lint distribution package setuptools installer.
```
Comment (webknjaz) [inline setup.py:2]: Not lint

File: setup.cfg [hunk 9]
```diff
@@ -0,0 +1,82 @@
+[aliases]
+dists = clean --all sdist bdist_wheel
+
```
Comment (webknjaz) [inline setup.cfg:3]: Nothing uses this

File: setup.cfg [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  CI: GitHub = https://github.com/ansible-community/ansible-compat/actions?query=workflow:gh+branch:master+event:push
+  Code of Conduct = https://docs.ansible.com/ansible/latest/community/code_of_conduct.html
+  Source Code = https://github.com/ansible-community/ansible-compat
+description = Ansible compatibility goodies
+long_description = file: README.rst
+long_description_content_type = text/x-rst
+author = Sorin Sbarnea
+author_email = ssbarnea@redhat.com
+maintainer = Sorin Sbarnea
+maintainer_email = ssbarnea@redhat.com
```
Comment (webknjaz) [inline setup.cfg:18]: I normally set the maintainer field if it's different from the author. Not sure if there's any policy wrt tho

File: setup.cfg [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  Operating System :: OS Independent
+
+  License :: OSI Approved :: MIT License
+  License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)
+
+  Programming Language :: Python
+  Programming Language :: Python :: 3
+  Programming Language :: Python :: 3.6
+  Programming Language :: Python :: 3.7
+  Programming Language :: Python :: 3.8
```
Comment (webknjaz) [inline setup.cfg:39]: 3.9?

File: setup.cfg [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  Programming Language :: Python :: Implementation :: PyPy
+
+  Topic :: Software Development :: Bug Tracking
+  Topic :: Software Development :: Quality Assurance
+  Topic :: Software Development :: Testing
+
+  Topic :: Utilities
+keywords =
+  ansible
+  lint
```
Comment (webknjaz) [inline setup.cfg:52]: not lint

File: setup.cfg [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+[options.extras_require]
+test =
+  pytest
+  pytest-cov
+
+[options.packages.find]
+where = src
+
+[options.package_data]
+ansiblelint = py.typed
```
Comment (webknjaz) [inline setup.cfg:79]: not lint

File: setup.cfg [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+[codespell]
+skip = .tox,.mypy_cache,build,.git,.eggs,pip-wheel-metadata
```
Comment (webknjaz) [inline setup.cfg:82]: Does it have its own config?

File: pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+build-backend = "setuptools.build_meta"
+
+[tool.black]
+skip-string-normalization = true
+
+[tool.isort]
+profile = "black"
+
+[tool.setuptools_scm]
+local_scheme = "no-local-version"
```
Comment (webknjaz) [inline pyproject.toml:17]: I'd add this in the CI only.

File: mypy.ini [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+disallow_untyped_calls = True
+disallow_untyped_defs = True
+disallow_any_generics = True
+; disallow_any_unimported = True
+; warn_redundant_casts = True
+; warn_return_any = True
+; warn_unused_configs = True
+exclude = test/local-content
+
+[mypy-pytest]
```
Comment (webknjaz) [inline mypy.ini:14]: I think the recent pytest has typing embedded.

File: constraints.txt [hunk 17]
```diff
@@ -0,0 +1,30 @@
+#
```
Comment (webknjaz) [inline constraints.txt:1]: I think I'd prefer this in a separate `requirements/` dir.

File: .yamllint [hunk 18]
```diff
@@ -0,0 +1,9 @@
+rules:
+  document-start: disable
+  indentation:
+    level: error
+    indent-sequences: consistent
+ignore: |
+  .tox
+  examples/playbooks/example.yml
```
Comment (webknjaz) [inline .yamllint:8]: This looks irrelevant.

File: .pre-commit-config.yaml [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+    # shell test does not seem to work in CIs:
+    entry: >-
+      sh -c 'git hash-object setup.py
+      |
+      python -c raise\ SystemExit\(input\(\)\ !=\ \"f6d1010b609cbe816d3ef652eee452d09d52979f\"\)
+      '
+    pass_filenames: false
+    language: system
+    files: >-
+      ^setup\.py$
```
Comment (webknjaz) [inline .pre-commit-config.yaml:20]: I'm hearing `pip` now special-cases setuptools to support editable installs w/o `setup.py`. Somebody needs to evaluate how well that works and maybe just wipe this file.

File: .pylintrc [hunk 20]
```diff
@@ -0,0 +1,32 @@
+[IMPORTS]
+preferred-modules =
+    unittest:pytest,
+
+[MESSAGES CONTROL]
+
+disable =
+    # On purpose disabled as we rely on black
+    line-too-long,
+    # TODO(ssbarnea): remove temporary skips adding during initial adoption:
```
Comment (webknjaz) [inline .pylintrc:10]: Looks like most of this should be dropped since it's a new project and it's best not to allow weird stuff from the very beginning.

File: .pre-commit-config.yaml [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+- repo: https://github.com/pre-commit/pre-commit-hooks.git
+  rev: v4.0.1
+  hooks:
+  - id: end-of-file-fixer
+  - id: trailing-whitespace
+    exclude: >
+      (?x)^(
+        examples/playbooks/(with-skip-tag-id|unicode).yml|
+        examples/playbooks/example.yml
+      )$
```
Comment (webknjaz) [inline .pre-commit-config.yaml:30]: irrelevant

File: .pre-commit-config.yaml [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+  - id: doc8
+- repo: https://github.com/adrienverge/yamllint.git
+  rev: v1.26.1
+  hooks:
+  - id: yamllint
+    exclude: >
+      (?x)^(
+        examples/playbooks/templates/.*|
+        examples/other/some.j2.yaml
+      )$
```
Comment (webknjaz) [inline .pre-commit-config.yaml:53]: irrelevant

File: .github/workflows/tox.yml [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+        - 3.9
+        # NOTE: Installing ansible under 3.10-dev is currently not
+        # NOTE: possible because compiling cffi explodes.
+        os:
+        # https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners
+        - ubuntu-20.04
+        # - windows-latest
+        # - windows-2016
+        include:
+        - tox_env: py36
```
Comment (webknjaz) [inline .github/workflows/tox.yml:117]: What's the point of having the toxenv for each interpreter if `tox -e py` picks up the current one already?

File: .github/workflows/release.yml [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+        --user
+        tox
+    - name: Check out src from Git
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 0  # needed by setuptools-scm
+    - name: Build dists
+      run: python -m tox
+    - name: Publish to test.pypi.org
+      if: >-  # "create" workflows run separately from "push" & "pull_request"
```
Comment (webknjaz) [inline .github/workflows/release.yml:37]: the comment doesn't match the content

File: tox.ini (hunk 3 above)
Comment (ssbarnea) [reply tox.ini:54]: This is included by default for packaging (inhereted).

File: constraints.txt (hunk 17 above)
Comment (ssbarnea) [reply constraints.txt:1]: I would like to hide many of the files from root under a subfolder but picking `requirements/` would not be a generic enough approach. Lets delay this until we can make a more generic solution, maybe under `.config/`?
